diff.go

```diff
@@ -847,3 +849,134 @@ func DiffBlobs(oldBlob *Blob, oldAsPath string, newBlob *Blob, newAsPath string,
 
 	return nil
 }
+
+type ApplyHunkCallback func(*DiffHunk) (apply bool, err error)
+type ApplyDeltaCallback func(*DiffDelta) (apply bool, err error)
+
+// ApplyOptions has 2 callbacks that are called for hunks or deltas
+// If these functions return an error, abort the apply process immediately.
+// If the first resutnr value is true, the delta/hunk will be applied. If it is false, the  delta/hunk will not be applied. In either case, the rest of the apply process will continue.
+type ApplyOptions struct {
+	Version uint
+	ApplyHunkCallback
+	ApplyDeltaCallback
+	Flags uint
+}
+
+//export hunkApplyCallback
+func hunkApplyCallback(_hunk *C.git_diff_hunk, _payload unsafe.Pointer) C.int {
+	opts, ok := pointerHandles.Get(_payload).(ApplyOptions)
+	if !ok {
+		panic("invalid apply options payload")
+	}
+
+	if opts.ApplyHunkCallback == nil {
+		return 0
+	}
+
+	hunk := diffHunkFromC(_hunk)
+
+	apply, err := opts.ApplyHunkCallback(&hunk)
+	if err != nil {
+		return -1
+	} else if apply {
+		return 0
+	} else {
+		return 1
+	}
+}
+
+//export deltaApplyCallback
+func deltaApplyCallback(_delta *C.git_diff_delta, _payload unsafe.Pointer) C.int {
+	opts, ok := pointerHandles.Get(_payload).(ApplyOptions)
+	if !ok {
+		panic("invalid apply options payload")
+	}
+
+	if opts.ApplyDeltaCallback == nil {
+		return 0
+	}
+
+	delta := diffDeltaFromC(_delta)
+
+	apply, err := opts.ApplyDeltaCallback(&delta)
+	if err != nil {
+		return -1
+	} else if apply {
+		return 0
+	} else {
+		return 1
+	}
+}
+
+func DefaultApplyOptions() (*ApplyOptions, error) {
+	opts := C.git_apply_options{}
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	C._go_git_apply_init_options(&opts)
```

Comment:
other places (like https://github.com/libgit2/git2go/blob/fc6eaf36388841b16ff004e1d48e887d3f9613dc/diff.go#L514) use the function initialization rather than the initialization constant to avoid having to create a wrapper:

```suggestion
	ecode := C.git_apply_options_init(&opts, C.GIT_APPLY_OPTIONS_VERSION)
```

Reply:
I thought there was a reason I did it this way but after changing it to call this it works fine :thinking: